flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -437,19 +440,20 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
 
 				wasEmpty = receivedBuffers.isEmpty();
 
-				if (buffer.getDataType().hasPriority()) {
-					receivedBuffers.addPriorityElement(new SequenceBuffer(buffer, sequenceNumber));
-					if (channelStatePersister.checkForBarrier(buffer)) {
-						// checkpoint was not yet started by task thread,
-						// so remember the numbers of buffers to spill for the time when it will be started
-						numBuffersOvertaken = receivedBuffers.getNumUnprioritizedElements();
-					}
-					firstPriorityEvent = receivedBuffers.getNumPriorityElements() == 1;
+				SequenceBuffer sequenceBuffer = new SequenceBuffer(buffer, sequenceNumber);
+				DataType dataType = buffer.getDataType();
+				if (dataType.hasPriority()) {
+					checkPriorityXorAnnouncement(buffer);
+					firstPriorityEvent = addPriorityBuffer(sequenceBuffer);
 				} else {
-					receivedBuffers.add(new SequenceBuffer(buffer, sequenceNumber));
+					receivedBuffers.add(sequenceBuffer);
 					channelStatePersister.maybePersist(buffer);
-				}
 
+					if (dataType.requiresAnnouncement()) {
+						checkPriorityXorAnnouncement(buffer);
+						firstPriorityEvent = addPriorityBuffer(announce(sequenceBuffer));
+					}
```

Comment:
```
if (dataType.hasPriority() || dataType.requiresAnnouncement()) {
  firstPriorityEvent = addPriorityBuffer(announce(sequenceBuffer));
}
if (!dataType.hasPriority()) {
  receivedBuffers.add(sequenceBuffer);
  channelStatePersister.maybePersist(buffer);
}
```

Reply:
This is incorrect, it would have to be:
```
				if (dataType.requiresAnnouncement()) {
					firstPriorityEvent = addPriorityBuffer(announce(sequenceBuffer));
				}
				if (dataType.hasPriority()) {
					firstPriorityEvent = addPriorityBuffer(sequenceBuffer);
				}
				else {
					receivedBuffers.add(sequenceBuffer);
					channelStatePersister.maybePersist(buffer);
				}
```
but that would suggest priority event can be announced, hence I moved 
```
				if (dataType.requiresAnnouncement()) {
					firstPriorityEvent = addPriorityBuffer(announce(sequenceBuffer));
				}
```
into the non priority branch.

Reply:
Missed the `announce` part.